### PR TITLE
Separate route calculating from network message handlers

### LINKED ISSUE #432: Split route calculation from channel rumor message handling
We currently have one object that both handles route calculation and receiving gossip messages (to create its routing db). We should split this in two so that its possible to have a remote server provide routing data while still calculating routes locally.

### PR DESCRIPTION
**Pull request by naumenkogs**
Closes #432 

~~Not fully ready yet, but it compiles and passes the tests, so I thought I could use some confirmation that I do sane things before I finish those small touches.~~

TODO left:
- [x] uncomment missing_docs and add/update docs for Router and NetworkGraphMonitor
- [x] double check that visibility (``pub``) is not too permissive

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
Nice! This generally looks good, though I'm not really a fan of exposing the graph structs as-is today (see conversation on #566) and feel pretty icky exposing a RwLock publicly, given its taken in our own handler stuff - this makes it somewhat easy to accidentally introduce deadlocks in user code by having the user hold a lock while handling some message which ends up taking the lock again.

Instead, for now, may be best to let the router take a reference to the NetworkGraph (I think you can drop the word Monitor) and then take a read lock in the get_route function.

**Comment by naumenkogs** [comment]
@TheBlueMatt Made some updates, should be ready for review!
I have couple questions though:

1. I’m not sure about just `NetworkGraph`name  because it has a field `NetworkMap` which to me sounds like a synonym, and because of that is a little confusing… “Monitor” would indicate that it’s something with more powers, won’t it?

2. It’s a little unclear where `mark_node_bad` should sit. I feel maybe it belongs to the Router, because it should be shared across different routing data providers.

3. Does `our_node_id` actually belong to NetworkMap? I feel it may be better in the Router.

4. What’s the purpose of `get_short_id` in NetworkMap? I see why it `get_key` makes sense when there are non-Bitcoin chains, but `get_short_id` is unclear. What’s the difference between them?

**Comment by naumenkogs** [comment]
Just saw those compiler issues on travis, will address soon. Questions above still apply :)

**Comment by naumenkogs** [comment]
Travis is happy now, will squash last commit when we agree on those questions above.

File: lightning/src/ln/network_graph.rs [hunk 1]
```diff
@@ -0,0 +1,1609 @@
...
+	logger: Arc<Logger>,
+}
+
+impl NetworkGraph {
+	/// Creates a new tracker of the actual state of the network of channels and nodes.
+	pub fn new(our_pubkey: PublicKey, chain_monitor: Arc<ChainWatchInterface>, logger: Arc<Logger>) -> Self {
+		let mut nodes = BTreeMap::new();
+		nodes.insert(our_pubkey.clone(), NodeInfo {
+			channels: Vec::new(),
+			lowest_inbound_channel_fee_base_msat: u32::max_value(),
```
Comment (TheBlueMatt) [inline lightning/src/ln/network_graph.rs:43]: If these are going to be pub we need something other than default values. We need them to actually be Option<>s and any groups of fields which are always either set or unset together should be in structs.

**Comment by naumenkogs** [comment]
Finished a bunch of work here, I didn't expect that many changes :)
I have some questions before the last push. 2 of them I leave as comments to the code on github, but one is here:

How do we see NetworkMap at the end?
I'm thinking merging it with NetworkGraph, but then, what happens to the RwLock<NetworkMap>?
Do we move the lock to nodes/channels fields within the resulting data structure? Or do we get rid of it?

File: lightning/src/routing/router.rs [hunk 2]
```diff
@@ -0,0 +1,1186 @@
...
+				let proportional_fee_millions = ($starting_fee_msat + final_value_msat).checked_mul($directional_info.fees.proportional_millionths as u64);
+				if let Some(new_fee) = proportional_fee_millions.and_then(|part| {
+						($directional_info.fees.base_msat as u64).checked_add(part / 1000000) })
+				{
+					let mut total_fee = $starting_fee_msat as u64;
+					let hm_entry = dist.entry(&$directional_info.src_node_id);
+					let old_entry = hm_entry.or_insert_with(|| {
+						let node = network.nodes.get(&$directional_info.src_node_id).unwrap();
+						let mut fee_base_msat = 0;
+						let mut fee_proportional_millionths = 0;
```
Comment (naumenkogs) [inline lightning/src/routing/router.rs:235]: We wanted to get rid of default values in node inbound fees, but this is the place about I left them.
I'm not sure I fully understand what's going on in this part of the code, so not sure this is correct.

File: lightning/src/routing/router.rs [hunk 3]
```diff
@@ -0,0 +1,1186 @@
...
+					None => return Err(LightningError{err: "Failed to find a non-fee-overflowing path to the given destination", action: ErrorAction::IgnoreError}),
+				};
+				res.last_mut().unwrap().fee_msat = new_entry.fee_msat;
+				res.last_mut().unwrap().cltv_expiry_delta = new_entry.cltv_expiry_delta;
+				res.push(new_entry);
+			}
+			res.last_mut().unwrap().fee_msat = final_value_msat;
+			res.last_mut().unwrap().cltv_expiry_delta = final_cltv;
+			let route = Route { paths: vec![res] };
+			// log_trace!(self, "Got route: {}", log_route!(route));
```
Comment (naumenkogs) [inline lightning/src/routing/router.rs:373]: I have no idea how this thing works and why it fails right now :)

File: lightning/src/routing/router.rs (hunk 2 above)
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs:235]: Right, sorry, lack of clarity on my end - I was dont like having default values in public interfaces (as Rust has a native Option<> for exactly this reason), but in this case were putting default values in a temporary map used for our routefinding, which is totally fine.

File: lightning/src/routing/router.rs (hunk 3 above)
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs:373]: our logging stuff generally assumes its being called from a method that takes an &self (ie a non-static method) and then accesses self.logger. While we should probably update our macros to make that not be the case, for now you can use the hack we have at https://github.com/rust-bitcoin/rust-lightning/blob/master/lightning/src/ln/onion_utils.rs#L443.

File: lightning/src/routing/network_graph.rs [hunk 4]
```diff
@@ -0,0 +1,1629 @@
...
+}
+
+/// Represents the network as nodes and channels between them
+#[derive(PartialEq)]
+pub struct NetworkMap {
+	/// Known valid channels
+	pub channels: BTreeMap<u64, ChannelInfo>,
+
+	/// Identifier of the local Lightning node
+	pub our_node_id: PublicKey,
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:356]: This should probably just a parameter to get_route.

File: lightning/src/routing/network_graph.rs [hunk 5]
```diff
@@ -0,0 +1,1629 @@
...
+			announcement_message
+		})
+	}
+}
+
+/// Represents the network as nodes and channels between them
+#[derive(PartialEq)]
+pub struct NetworkMap {
+	/// Known valid channels
+	pub channels: BTreeMap<u64, ChannelInfo>,
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:353]: I dont know how I feel about these being pub: RW. We have a few consistency requirements on this and nodes, etc, so maybe it makes more sense to just have accessors for them and options to add new channels directly (without going through the message abstractions)?

File: lightning/src/routing/router.rs (hunk 2 above)
Comment (naumenkogs) [reply lightning/src/routing/router.rs:235]: Right, I guess that's how the code should be read.

**Comment by naumenkogs** [comment]
@TheBlueMatt Okay, just wanted to clarify before finishing the last commit.
What you are suggesting is:

```struct NetworkGraph (ex-NetworkMap) {
     (not pub, no lock) channels;
     (not pub, no lock) nodes
} + function like AddChannel() and GetNodes()
```

```struct NetGraphMessageHandler {
     pub RWLock<NetworkGraph> network_graph;
....
}
```
and get_route in this case takes a reference to NetGraphMessageHandler? 

I feel a bit odd to pass something message handler to be used for routing.

Alternative (similarly to what happens with channels) would be to call that message handler NetworkGraphManager.

**Comment by TheBlueMatt** [comment]
You could make the internal fields read-only with something like this:

```
struct NetworkGraph (ex-NetworkMap) {
     (not pub, no lock) channels: HashMap<...>
     (not pub, no lock) nodes
} + function like AddChannel()
impl NetworkGraph {
    // TODO: AddChannel function
    fn get_channels<'a>(&'a self) -> &'a HashMap<...> { &self.channels }
```

**Comment by naumenkogs** [comment]
Alright, ready for review!

There's some non-necessary renames (like NetworkGraphMonitor->NetworkGraph->NetGraphMsgHandler) in the middle of the PR, but they would be really difficult to squash/reorg at this point, so maybe we can keep commits as is?

Also, I'm still a little unsure about our_node in NetGraphMsgHandler, now in the constructor. Should we keep it there? Or maybe keep, but make it optional.

**Comment by TheBlueMatt** [comment]
> Also, I'm still a little unsure about our_node in NetGraphMsgHandler, now in the constructor. Should we keep it there? Or maybe keep, but make it optional.

I thought we were just going to make that a parameter to get_route?

File: lightning/src/routing/network_graph.rs [hunk 6]
```diff
@@ -0,0 +1,1640 @@
...
+				} else {
+					panic!("Had channel that pointed to unknown node (ie inconsistent network map)!");
+				}
+			}
+		}
+		remove_from_node!(chan.one_to_two.src_node_id);
+		remove_from_node!(chan.two_to_one.src_node_id);
+	}
+
+	/// Marks a node as having failed a route. This will avoid re-using the node in routes for now,
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:88]: Honestly, we should probably just wholesale remove this - its not implemented and the API doesn't even make sense...we need a much more wholesale routing state machine.

File: lightning/src/routing/network_graph.rs [hunk 7]
```diff
@@ -0,0 +1,1640 @@
...
+		for (key, val) in self.nodes.iter() {
+			write!(f, " {}: {}\n", log_pubkey!(key), val)?;
+		}
+		Ok(())
+	}
+}
+
+impl NetworkGraph {
+	#[inline]
+	/// Key used to link channel on the Bitcoin chain to other entities
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:430]: We can drop these - they were there because we had different map keys depending on the other_chains compile flag, but that is gone now.

File: lightning/src/routing/network_graph.rs [hunk 8]
```diff
@@ -0,0 +1,1640 @@
...
+	}
+
+	#[inline]
+	/// Key used to link channel on the Bitcoin chain to other entities
+	pub fn get_short_id(id: &u64) -> &u64 {
+		id
+	}
+}
+
+impl RoutingMessageHandler for NetGraphMsgHandler {
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:442]: Hmm, can you move this up in the file so all the NetGraphMsgHandler stuff is together and all the NetworkGraph stuff is together?

**Comment by naumenkogs** [comment]
>I thought we were just going to make that a parameter to get_route?

Yeah, it is, with the current code. Since it's an explicit parameter, I assume it is no longer needed in NetGraphMsgHandler (it also doesn't make much sense).

But currently it is still passed to the constructor of NetGraphMsgHandler, just to add `our` node to the graph. So my question is about this.

File: lightning/src/routing/network_graph.rs [hunk 9]
```diff
@@ -0,0 +1,1640 @@
...
+
+	fn handle_node_announcement(&self, msg: &msgs::NodeAnnouncement) -> Result<bool, LightningError> {
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:444]: I feel like if we probably want these to be in NetworkGraph, not NetGraphMsgHandler and just have thing wrappers across NetGraphMsgHandler. At a high level, the goal of encapsulating NetworkGraph and not letting users simply have direct write access to all the internal fields is that they have to stay consistent - by allowing someone to, eg, take a NodeAnnouncement, pass it into a NetworkGraph (with &mut self), we can easily provide the ability to update NetworkGraph without leaving us in an inconsistent state.

File: lightning/src/routing/network_graph.rs [hunk 10]
```diff
@@ -0,0 +1,1640 @@
...
+	channels: BTreeMap<u64, ChannelInfo>,
+	/// A set of known Lightning nodes and their properties
+	nodes: BTreeMap<PublicKey, NodeInfo>,
+}
+
+impl NetworkGraph {
+	/// Returns a list of known valid channels
+	pub fn get_channels<'a>(&'a self) -> &'a BTreeMap<u64, ChannelInfo> { &self.channels }
+	/// Returns a mutable reference to an existing channel
+	pub fn get_channel_mut<'a>(&'a mut self, chan_id: &u64) -> Option<&mut ChannelInfo> { self.channels.get_mut(chan_id) }
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:360]: We can't just expose these completely writeable, we want to allow users to add channels/nodes, but only in a way that keeps the internal state consistent. Otherwise we'll end up with panics like the one you showed at https://github.com/rust-bitcoin/rust-lightning/pull/570#discussion_r403817262.

**Comment by TheBlueMatt** [comment]
> But currently it is still passed to the constructor of NetGraphMsgHandler, just to add our node to the graph. So my question is about this.

Ah, right, yea, I think we can drop that - tests pass fine without it mod a few routing tests that assume its there, which implies we're still able to route just fine (as I would expect).

**Comment by naumenkogs** [comment]
Oh no, now there are conflicts in a gazillion of tests again. Will rebase tomorrow I guess.

There are more commits than it should be, but selective squashing at this point would be painful, mostly because of the tests refactoring… Like every time I change the netgraphwhatever interface, 40 tests has to be updated.
I probably can squash all them into 3 commits in total if that helps.

Last thing would be to update ARCH.md, once this passes the review.

**Comment by naumenkogs** [comment]
@TheBlueMatt rebasing was painful, so I had to squash everything and split into 3 commits.
Can split further, if that would really help.
The biggest problem is that any further split requires to touch those many-many tests.

File: lightning/src/routing/network_graph.rs [hunk 11]
```diff
@@ -0,0 +1,1596 @@
...
+		let network = self.network_graph.read().unwrap();
+		network.get_nodes().get(pubkey).map(|n| n.addresses.clone())
+	}
+
+	/// Dumps the entire network view of this NetGraphMsgHandler to the logger provided in the constructor at
+	/// level Trace
+	pub fn trace_state(&self) {
+		log_trace!(self, "{}", self.network_graph.read().unwrap());
+	}
+
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:62]: nit: careful with excess whitespace in a number of places.

File: lightning/src/routing/network_graph.rs [hunk 12]
```diff
@@ -0,0 +1,1596 @@
...
+		let network = self.network_graph.read().unwrap();
+		network.write(writer)?;
+		Ok(())
+	}
+}
+
+/// Arguments for the creation of a NetGraphMsgHandler that are not deserialized.
+/// At a high-level, the process for deserializing a NetGraphMsgHandler and resuming normal operation is:
+/// 1) Deserialize the NetGraphMsgHandler by filling in this struct and calling <NetGraphMsgHandler>::read(reaser, args).
+/// 2) Register the new NetGraphMsgHandler with your ChainWatchInterface
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:214]: I dont think you need to do this - we only use the chain_monitor for queries, the NetGraphMsgHandler doesn't (currently) have any need for receiving block updates. Maybe in the future we can do that but its not implemented yet.

File: lightning/src/routing/network_graph.rs [hunk 13]
```diff
@@ -0,0 +1,1596 @@
...
+	last_update_message
+});
+
+#[derive(PartialEq)]
+/// Details regarding a channel (both directions)
+pub struct ChannelInfo {
+	/// Protocol features of a channel communicated during its announcement
+	pub features: ChannelFeatures,
+	/// Details regarding one of the directions of a channel
+	pub one_to_two: DirectionalChannelInfo,
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:285]: Shouldn't these be Option<>al? We don't set them until later, that way we could avoid default-value-filling them.

File: lightning/src/routing/network_graph.rs [hunk 14]
```diff
@@ -0,0 +1,1596 @@
...
+	/// Details regarding another direction of a channel
+	pub two_to_one: DirectionalChannelInfo,
+	/// An initial announcement of the channel
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:288]: Can you describe when this isn't filled in and why its Option<>al (namely that the data is redundant and we sometimes avoid storing things here if announcements include lots of extra data that we don't know how to parse, avoiding some basic DoS attacks).

File: lightning/src/routing/network_graph.rs [hunk 15]
```diff
@@ -0,0 +1,1596 @@
...
+	}
+}
+
+
+#[derive(PartialEq)]
+/// Details regarding a node in the network
+pub struct NodeInfo {
+	/// All valid channels a node has announced
+	pub channels: Vec<u64>,
+	/// Lowest fees enabling routing via any of the known channels to a node
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:343]: Can you describe when its None?

File: lightning/src/routing/network_graph.rs [hunk 16]
```diff
@@ -0,0 +1,1596 @@
...
+	pub lowest_inbound_channel_fees: Option<RoutingFees>,
+	/// Protocol features the node announced support for
+	pub features: NodeFeatures,
+	/// When the last known update to the node state was issued
+	/// Unlike for channels, we may have a NodeInfo entry before having received a node_update.
+	/// Thus, we have to be able to capture "no update has been received", which we do with an
+	/// Option here.
+	pub last_update: Option<u32>,
+	/// Color assigned to the node
+	pub rgb: [u8; 3],
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:353]: Why arent these Options? Can we refactor all the stuff that shows up in the NodeAnnouncement into a separate struct and make it Option<>al here?

File: lightning/src/routing/network_graph.rs [hunk 17]
```diff
@@ -0,0 +1,1596 @@
...
+		write!(f, "[Nodes]\n")?;
+		for (key, val) in self.nodes.iter() {
+			write!(f, " {}: {}\n", log_pubkey!(key), val)?;
+		}
+		Ok(())
+	}
+}
+
+impl NetworkGraph {
+	/// Returns a list of known valid channels
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:490]: Its not a list, its a BTree :p (from short channel id -> ChannelInfo).

File: lightning/src/routing/network_graph.rs [hunk 18]
```diff
@@ -0,0 +1,1596 @@
...
+	pub fn get_channels<'a>(&'a self) -> &'a BTreeMap<u64, ChannelInfo> { &self.channels }
+	/// Returns a list of known nodes
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:492]: Same, its not a list.

File: lightning/src/routing/network_graph.rs [hunk 19]
```diff
@@ -0,0 +1,1596 @@
...
+	pub fn get_nodes<'a>(&'a self) -> &'a BTreeMap<PublicKey, NodeInfo> { &self.nodes }
+
+	fn process_node_announcement(&mut self, msg: &msgs::NodeAnnouncement) -> Result<bool, LightningError> {
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:495]: I think these should be pub. Its fine to do it in a separate PR, but they're pretty useful, at least once they have sufficient documentation. May also make sense to not call them "process" as much as add_node_from_announcement or update_node_from_announcement or something like that, making sure to note in the docs that signatures are *not* checked.

Comment (naumenkogs) [reply lightning/src/routing/network_graph.rs:495]: I will do this, but if we making them public, does it make sense to move should_relay to the message handler, one level above?
It probably doesn't make much sense for every caller to know that, and if they want to relay or not, they can check it on their own, since it's trivial now and won't need db access.

File: lightning/src/routing/network_graph.rs (hunk 16 above)
Comment (naumenkogs) [reply lightning/src/routing/network_graph.rs:353]: Doing this for rgb and alias, not sure about features and addresses?

Because when we interpret features/addrs for None, we will indeed return Features::empty and empty vector, not just because it's a default value, but because that's the correct answer for a given call.

File: lightning/src/routing/network_graph.rs (hunk 11 above)
Comment (naumenkogs) [reply lightning/src/routing/network_graph.rs:62]: found two, i hope that was it..

File: lightning/src/routing/network_graph.rs (hunk 12 above)
Comment (naumenkogs) [reply lightning/src/routing/network_graph.rs:214]: Does it even make sense to serialize NetGraphMsgHandler at all? The only field there is NetworkGraph, which can be serialized on it's own, and then spawn a new MsgHandler when needed.

File: lightning/src/routing/network_graph.rs (hunk 13 above)
Comment (naumenkogs) [reply lightning/src/routing/network_graph.rs]: Almost finished this.
The last bit I'm not sure about is handling last hops. Route_test currently fails, because it passes unknown channels for last hops, and [I assume](https://github.com/rust-bitcoin/rust-lightning/blob/a62f10676b00f5a9816d87b0960099992affe276/lightning/src/routing/router.rs#L341) they are known.

I could have added announcement for them before the test, but I wasn't sure what's fair to expect from user here.

Comment (TheBlueMatt) [reply lightning/src/routing/network_graph.rs]: No, yea, the expectation is that if there is something in the last hops set, we ignore the channel graph for those and pull from the last hops set, since those should be the most up-to-date (they were included in the invoice).

File: lightning/src/routing/network_graph.rs (hunk 16 above)
Comment (TheBlueMatt) [reply lightning/src/routing/network_graph.rs]: For addresses, yea, I dont think so for features. There is  big difference between "dont know their feature set" and "they don't support anything".

Comment (TheBlueMatt) [reply lightning/src/routing/network_graph.rs]: Oh, but actually, why not just make it one struct. This is a pretty classic example of "datastructures should capture what the reality of possible states are" - rgb, alias, addresses, features, and announcement_message should be in a separate sub-struct, eg the following. This super neatly captures, in both a human-understandable, as well as machine-understandable (ie compiler-enforced) way, that this set of fields is available only after the first announcement info, and after that, they are all available.

```
/// Information received in the latest node_announcement from this node.
pub struct NodeAnnouncementInfo {
       pub features: NodeFeatures,
       pub last_update: u32,
       pub rgb:[u8; 3],
       pub alias: [u8; 32],
       pub addresses: Vec<NetAddress>,
       pub announcement_message: Option<msgs::NodeAnnouncement>,
}
pub struct NodeInfo {
       pub channels: Vec<u64>,
       pub lowest_inbound_channel_fees: Option<RoutingFees>,
       pub announcement_info: Option<NodeAnnouncementInfo>
}
```

**Comment by naumenkogs** [comment]
Addressed all the pending feedback. Travis passed, not sure why it shows waiting here.

File: lightning/src/routing/network_graph.rs [hunk 20]
```diff
@@ -0,0 +1,1672 @@
...
+	fees,
+	last_update_message
+});
+
+#[derive(PartialEq)]
+/// Details regarding a channel (both directions)
+pub struct ChannelInfo {
+	/// Protocol features of a channel communicated during its announcement
+	pub features: ChannelFeatures,
+	/// Source node of one of the directions of a channel
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:294]: Maybe rename the variable to just be node_one and then its clear which direction goes with which node?

File: lightning/src/routing/network_graph.rs (hunk 14 above)
Comment (TheBlueMatt) [reply lightning/src/routing/network_graph.rs:302]: "//" is a comment which is not published in the documentation, you have to use "///" for that. It would be nice, still, to describe why it might be None (ie if there is extra data and we dont want to store it). Not sure why hit was marked resolved.

File: lightning/src/routing/network_graph.rs [hunk 21]
```diff
@@ -0,0 +1,1672 @@
...
+	announcement_message
+});
+
+
+/// Fees for routing via a given channel or a node
+#[derive(Eq, PartialEq, Copy, Clone, Debug)]
+pub struct RoutingFees {
+	/// Flat routing fee
+	pub base_msat: u32,
+	/// Liquidity-based routing fee
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:331]: Please specify units.

File: lightning/src/routing/network_graph.rs [hunk 22]
```diff
@@ -0,0 +1,1672 @@
...
+#[derive(PartialEq, Debug)]
+/// Information received in the latest node_announcement from this node.
+pub struct NodeAnnouncementInfo {
+	/// Protocol features the node announced support for
+      pub features: NodeFeatures,
+	/// When the last known update to the node state was issued
+      pub last_update: u32,
+	/// Color assigned to the node
+	pub rgb: [u8; 3],
+	/// Moniker assigned to the node
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:363]: Some details about this possibly being invalid or malicious (eg control chars) would be good. Probably worth suggesting this not be exposed to users.

File: lightning/src/routing/network_graph.rs [hunk 23]
```diff
@@ -0,0 +1,1672 @@
...
+	pub alias: [u8; 32],
+	/// Internet-level addresses via which one can connect to the node
+	pub addresses: Vec<NetAddress>,
+	/// An initial announcement of the node
+	// this is cached here so we can send out it later if required by initial routing sync
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:368]: Same here. Please provide further docs, not just a comment.

File: lightning/src/routing/network_graph.rs [hunk 24]
```diff
@@ -0,0 +1,1672 @@
...
+
+#[derive(PartialEq)]
+/// Details regarding a node in the network
+pub struct NodeInfo {
+	/// All valid channels a node has announced
+	pub channels: Vec<u64>,
+	/// Lowest fees enabling routing via any of the known channels to a node
+	pub lowest_inbound_channel_fees: Option<RoutingFees>,
+	/// More information about a node from node_announcement
+	/// Optional because we may have a NodeInfo entry before having received the announcement
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:424]: *do* have :p. Its actually illegal to receive node_announcement before having received channels for the node (which will create this).

File: lightning/src/routing/network_graph.rs [hunk 25]
```diff
@@ -0,0 +1,1672 @@
...
+		write!(f, "[Nodes]\n")?;
+		for (key, val) in self.nodes.iter() {
+			write!(f, " {}: {}\n", log_pubkey!(key), val)?;
+		}
+		Ok(())
+	}
+}
+
+impl NetworkGraph {
+	/// Returns all known valid channels
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:528]: Please specify the return types.

File: lightning/src/routing/network_graph.rs [hunk 26]
```diff
@@ -0,0 +1,1672 @@
...
+	pub fn get_channels<'a>(&'a self) -> &'a BTreeMap<u64, ChannelInfo> { &self.channels }
+	/// Returns all known nodes
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:530]: Same.

File: lightning/src/routing/network_graph.rs [hunk 27]
```diff
@@ -0,0 +1,1672 @@
...
+					announcement_message: if should_relay { Some(msg.clone()) } else { None },
+				});
+
+				Ok(should_relay)
+			}
+		}
+	}
+
+	/// For a new or already known (from previous announcement) channel, store or update channel info,
+	/// after making sure it corresponds to a real transaction on-chain.
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:561]: We dont do this? Thats what checked_utxo is for? checked_utxo's properties should be detailed, however.

File: lightning/src/routing/network_graph.rs [hunk 28]
```diff
@@ -0,0 +1,1672 @@
...
+	fn fail_node(&mut self, _node_id: &PublicKey, is_permanent: &bool) {
+		if *is_permanent {
+			// TODO: Wholly remove the node
+		} else {
+			// TODO: downgrade the node
+		}
+	}
+
+	/// For an already known (from announcement) channel, update info regarding one of the directions of a channel.
+	/// Announcement signatures *are checked* here and should be checked by a caller beforehand.
```
Comment (TheBlueMatt) [inline lightning/src/routing/network_graph.rs:654]: Why should they be checked beforehand if they're checked here? :p. I get the reasoning (the pubkey isnt available) but its super awkward to have it be different across different fns. Maybe we should just always do sig checking in NetworkGraph?

File: lightning/src/routing/network_graph.rs (hunk 25 above)
Comment (naumenkogs) [reply lightning/src/routing/network_graph.rs:528]: I don't understand this one... Docs usually capture return values pretty [well](https://docs.rs/lightning/0.0.11/lightning/ln/router/struct.Router.html#method.get_addresses) automatically. What do you think is unclear here?

File: lightning/src/routing/network_graph.rs [hunk 29]
```diff
@@ -0,0 +1,1672 @@
...
+	}
+}
+
+#[derive(PartialEq)]
+/// Details regarding a node in the network
+pub struct NodeInfo {
+	/// All valid channels a node has announced
+	pub channels: Vec<u64>,
+	/// Lowest fees enabling routing via any of the known channels to a node
+	pub lowest_inbound_channel_fees: Option<RoutingFees>,
```
[the remark this answers is not in the document]
Comment (naumenkogs) [reply lightning/src/routing/network_graph.rs:422]: Agreed it is confusing, but I don't think splitting fields would help much... We have to still apply the same comments and expect the user to read them.
So far planning to add this:
	/// The two fields (flat and proportional fee) are independent,
	/// meaning they don't have to refer to the same channel.
Don't know it's "greatly expanded", but not sure what else to say here.

File: lightning/src/routing/network_graph.rs (hunk 14 above)
Comment (naumenkogs) [reply lightning/src/routing/network_graph.rs:302]: It was resolved because I did add this comment, and then it got lost somewhere :(

File: lightning/src/routing/network_graph.rs (hunk 12 above)
Comment (naumenkogs) [reply lightning/src/routing/network_graph.rs:227]: You mean a NetworkGraph?
Not sure why the whole NetGraphMsgHandler would be useful...

File: lightning/src/routing/network_graph.rs (hunk 28 above)
Comment (naumenkogs) [reply lightning/src/routing/network_graph.rs:654]: But then in handle_channel_announcement we will be looking into the chain before checking sigs...
I guess it doesn't change much in this context.

File: lightning/src/routing/network_graph.rs (hunk 25 above)
Comment (TheBlueMatt) [reply lightning/src/routing/network_graph.rs:494]: What does the u64 in the hashmap represent, mostly.

File: lightning/src/routing/network_graph.rs (hunk 28 above)
Comment (TheBlueMatt) [reply lightning/src/routing/network_graph.rs]: Hmm, ugh, good point. There's no winning here. Maybe just make the signature checking optional with both variants exposed for each and use the relevant one from the message handler?

File: lightning/src/ln/functional_test_utils.rs [hunk 30]
```diff
@@ -97,36 +99,6 @@ impl<'a, 'b, 'c> Drop for Node<'a, 'b, 'c> {
 			assert!(self.node.get_and_clear_pending_events().is_empty());
 			assert!(self.chan_monitor.added_monitors.lock().unwrap().is_empty());
 
-			// Check that if we serialize the Router, we can deserialize it again.
```
Comment (TheBlueMatt) [inline lightning/src/ln/functional_test_utils.rs:100]: Why did this get removed?

File: lightning/src/routing/network_graph.rs (hunk 28 above)
Comment (naumenkogs) [reply lightning/src/routing/network_graph.rs]: Technically, it's enough in this case to just make secp_ctx optional.

But maybe we want to force users to explicitly set `check_sig` variable, and then if secp_ctx is None, throw an exception?

Not sure which approach we should take within this project.

File: lightning/src/ln/functional_test_utils.rs (hunk 30 above)
Comment (naumenkogs) [reply lightning/src/ln/functional_test_utils.rs:100]: Network graph serialization is already tested [here](https://github.com/naumenkogs/rust-lightning/blob/97ac065418832b7e0e42de246d7606241a235ae4/lightning/src/routing/network_graph.rs#L1584), unless you mean something else.

Comment (TheBlueMatt) [reply lightning/src/ln/functional_test_utils.rs:100]: This check is a bit different - in the past we've had issues in edge cases failing to serialize-deserialize properly. So we added this so that after every test (which often test fun edge cases), we can serialize-deserialize everything properly. I think we should keep this, at least for the underlying graph.

File: lightning/src/routing/network_graph.rs (hunk 28 above)
Comment (TheBlueMatt) [reply lightning/src/routing/network_graph.rs]: Yea, I think provide-a-context-if-you-want-signatures-checked is pretty nice. just pass in an Option<&Secp256k1> and document it. No need to panic or force users to set a variable. good docs and a clear None/Some() distinction requires typing and thinking so I'm not too worried about it.

**Comment by naumenkogs** [comment]
Addressed the last comments: returned serialization and updated Network Graph API.

**Comment by naumenkogs** [comment]
Not sure what i'm supposed to do with `codecov`...

**Comment by TheBlueMatt** [comment]
Just a few nits left. I added fixes for a few things at https://github.com/TheBlueMatt/rust-lightning/commits/2020-05-592-nits. I'm happy to just merge that branch, or you're welcome to squash them down onto the relevant commits here.

That said, before merge, a few of your changes should be squashed down (eg the fuzz fixes, etc) - all tests (including cd fuzz && cargo test) should build and pass after each individual commit.

**Comment by naumenkogs** [comment]
@TheBlueMatt I squashed fuzzing into commits so that every single commit passes tests.
Not sure what else you wanted to squash, if any?

If none, feel free to merge your branch now. I reviewed your code. looks good.

**Comment by TheBlueMatt** [comment]
Will take as #621.
